qa/tasks/keystone.py

```diff
@@ -0,0 +1,405 @@
+                     cconfig['roles'])
+    run_section_cmds(ctx, cclient, 'role add', 'name',
+                     cconfig['role-mappings'])
+    run_section_cmds(ctx, cclient, 'service create', 'name',
+                     cconfig['services'])
+    run_section_cmds(ctx, cclient, 'endpoint create', 'service',
+                     cconfig['endpoints'])
+
+    # sleep driven synchronization -- just in case
+    run_in_keystone_venv(ctx, cclient, [ 'sleep', '3' ])
+    try:
+        yield
+    finally:
+        pass
+
+
+@contextlib.contextmanager
+def task(ctx, config):
+    """
+    Deploy and configure Keystone
+
+    Example of configuration:
+
+      tasks:
+      - local_cluster:
+          cluster_path: /home/rzarzynski/ceph-1/build
+      - local_rgw:
+      - keystone:
+          client.0:
+            force-branch: master
+            tenants:
+              - name: admin
+                description:  Admin Tenant
+              - name: tempest
+                description: Tempest Tenant
+            users:
+              - name: admin
+                password: ADMIN
+                project: admin
+              - name: tempest_u1
+                password: tempest
+                project: tempest
+              - name: tempest_u2
+                password: tempest
+                project: tempest
+            roles: [ name: admin, name: Member ]
+            role-mappings:
+              - name: admin
+                user: admin
+                project: admin
+              - name: Member
+                user: tempest_u1
+                project: tempest
+              - name: Member
+                user: tempest_u2
+                project: tempest
+            services:
+              - name: keystone
+                type: identity
+                description: Keystone Identity Service
+              - name: swift
+                type: object-store
+                description: Swift Service
+              endpoints:
+                - service: keystone
+                  publicurl: http://127.0.1:$(public_port)s/v2.0
+                - service: swift
+                  publicurl: http://127.0.0.1:8000/v1/KEY_$(tenant_id)s
+    """
+    assert config is None or isinstance(config, list) \
+        or isinstance(config, dict), \
+        "task keystone only supports a list or dictionary for configuration"
+    all_clients = ['client.{id}'.format(id=id_)
+                   for id_ in teuthology.all_roles_of_type(ctx.cluster, 'client')]
+    if config is None:
+        config = all_clients
+    if isinstance(config, list):
+        config = dict.fromkeys(config)
+    clients = config.keys()
+
+    overrides = ctx.config.get('overrides', {})
+    # merge each client section, not the top level.
+    for client in config.iterkeys():
+        if not config[client]:
+            config[client] = {}
+        teuthology.deep_merge(config[client], overrides.get('keystone', {}))
```

Comment:
why this special handling of overrides? if you have multiple instances, you may want to apply different overrides to each